Index: target/linux/ramips/dts/mt7620n.dtsi
===================================================================
--- target/linux/ramips/dts/mt7620n.dtsi	(revision 42231)
+++ target/linux/ramips/dts/mt7620n.dtsi	(working copy)
@@ -204,6 +204,12 @@
 		pinctrl-0 = <&state_default>;
 		state_default: pinctrl0 {
 		};
+		i2c_pins: i2c {
+			i2c {
+				ralink,group = "i2c";
+				ralink,function = "i2c";
+			};
+		};
 		spi_pins: spi {
 			spi {
 				ralink,group = "spi";
@@ -210,6 +216,12 @@
 				ralink,function = "spi";
 			};
 		};
+		uartf_pins: uartf {
+			uartf {
+				ralink,group = "uartf";
+				ralink,function = "uartf";
+			};
+		};
 		uartlite_pins: uartlite {
 			uart {
 				ralink,group = "uartlite";
@@ -216,6 +228,36 @@
 				ralink,function = "uartlite";
 			};
 		};
+		wdt_pins: wdt {
+			wdt {
+				ralink,group = "wdt";
+				ralink,function = "wdt";
+			};
+		};
+		pa_pins: pa {
+			pa {
+				ralink,group = "pa";
+				ralink,function = "pa";
+			};
+		};
+		mdio_pins: mdio {
+			mdio {
+				ralink,group = "mdio";
+				ralink,function = "mdio";
+			};
+		};
+		rgmii1_pins: rgmii1 {
+			rgmii1 {
+				ralink,group = "rgmii1";
+				ralink,function = "rgmii1";
+			};
+		};
+		phy_led_pins: phy_led {
+			phy_led {
+				ralink,group = "led";
+				ralink,function = "led";
+			};
+		};
 	};
 
 	rstctrl: rstctrl {
